tatanka/client/conn/conn.go

```diff
@@ -43,79 +47,119 @@ type tatanka struct {
-		if err != nil {
-			return nil, err
-		}
-		b = append(b, block...)
+	id                 tanka.PeerID
+	ephemeralSharedKey []byte
+}
+
+const (
+	aesKeySize   = 32
+	gcmNonceSize = 12
+)
+
+func decryptAES(key []byte, ciphertext []byte) ([]byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create AES cipher: %v", err)
 	}
-	return b, nil
+
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create GCM: %v", err)
+	}
+
+	if len(ciphertext) < gcmNonceSize {
+		return nil, fmt.Errorf("ciphertext too short")
+	}
+	nonce, ciphertext := ciphertext[:gcmNonceSize], ciphertext[gcmNonceSize:]
+
+	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
+	if err != nil {
+		return nil, fmt.Errorf("decryption failed: %v", err)
+	}
+
+	return plaintext, nil
 }
 
-func (p *peer) encryptRSA(b []byte) ([]byte, error) {
-	msgLen := len(b)
-	hasher := blake256.New()
-	step := p.encryptionKey.Size() - 2*hasher.Size() - 2
-	var enc []byte
-	for start := 0; start < msgLen; start += step {
-		finish := start + step
-		if finish > msgLen {
-			finish = msgLen
-		}
-		block, err := rsa.EncryptOAEP(hasher, rand.Reader, p.encryptionKey, b[start:finish], []byte{})
-		if err != nil {
-			return nil, err
-		}
-		enc = append(enc, block...)
+func decryptTankagramPayload(key []byte, ciphertext []byte) (*msgjson.Message, error) {
+	plaintext, err := decryptAES(key, ciphertext)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt payload: %v", err)
+	}
+
+	var msg msgjson.Message
+	if err := json.Unmarshal(plaintext, &msg); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal payload: %v", err)
 	}
-	return enc, nil
+
+	return &msg, nil
 }
 
-// IncomingPeerConnect will be emitted when a peer requests a connection for
-// the transmission of tankagrams.
-type IncomingPeerConnect struct {
-	PeerID tanka.PeerID
-	Reject func()
+func decryptTankagramResult(key []byte, ciphertext []byte) (*mj.TankagramResultPayload, error) {
+	fmt.Printf("decrypting tankagram result with key %s\n", hex.EncodeToString(key))
+
+	plaintext, err := decryptAES(key, ciphertext)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt payload: %v", err)
+	}
+
+	var res mj.TankagramResultPayload
+	if err := json.Unmarshal(plaintext, &res); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal payload: %v", err)
+	}
+
+	return &res, nil
+}
+
+func encryptAES(key []byte, plaintext []byte) ([]byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create AES cipher: %v", err)
+	}
+
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create GCM: %v", err)
+	}
+
+	nonce := make([]byte, gcmNonceSize)
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, fmt.Errorf("failed to generate nonce: %v", err)
+	}
+
+	ciphertext := gcm.Seal(nonce, nonce, plaintext, nil)
```

Comment:
Is this always as long as `gcmNonceSize`? Otherwise will error in decrypt. I guess would need to pad `plaintext` if not.

Reply:
What do you mean? The nonce will always be the first 12 bytes of the ciphertext.

Reply:
In decryptAES there is:

```
	if len(ciphertext) < gcmNonceSize {
		return nil, fmt.Errorf("ciphertext too short")
	}
```

So my question is if the ciphertext is always long enough.

Reply:
The total length of `ciphertext` returned from `gcm.Seal` is `nonce size + plaintext size + tag size` so it’s always long enough, even if the plaintext is empty.